File: backend/src/main/java/cn/etarch/mao/harness/core/TokenEstimator.java
package cn.etarch.mao.harness.core;

import cn.etarch.mao.harness.llm.ChatRequest;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.nio.charset.StandardCharsets;
import java.util.List;

/**
 * 基于 UTF-8 字节数的粗算 Token 估算器：{@code ceil(bytes / 4)}。
 * 结构化内容先序列化为 JSON 再按可见字节估算；图片使用定额，不对 base64 做字节/4。
 */
@Slf4j
@Component
public class TokenEstimator {

    private final ObjectMapper objectMapper;

    // OpenAI chat format overhead per message
    private static final int MESSAGE_OVERHEAD_TOKENS = 4; // <|start|>{role}\n...\n<|end|>\n

    public TokenEstimator(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
        log.info("TokenEstimator initialized with UTF-8 bytes/4 heuristic");
    }

    /**
     * 估算完整 ChatRequest 的 token 数（含 system prompt + messages + tool definitions）
     */
    public int estimateRequestTokens(ChatRequest request) {
        int total = 0;

        // Messages
        if (request.getMessages() != null) {
            total += estimateMessages(request.getMessages());
        }

        // Tool definitions
        if (request.getTools() != null && !request.getTools().isEmpty()) {
            total += estimateToolDefinitions(request.getTools());
        }

        return total;
    }

    /**
     * 估算消息列表的 token 数
     */
    public int estimateMessages(List<ChatRequest.Message> messages) {
        int total = 0;
        for (ChatRequest.Message msg : messages) {
            total += estimateMessage(msg);
        }
        return total;
    }

    /**
     * 估算单条消息的 token 数
     */
    public int estimateMessage(ChatRequest.Message message) {
        int tokens = MESSAGE_OVERHEAD_TOKENS;

        // Role name tokens
        if (message.getRole() != null) {
            tokens += countTokens(message.getRole());
        }

        // Content tokens
        if (message.getContent() != null) {
            tokens += countTokens(contentToString(message.getContent()));
            tokens += countImagePartTokens(message.getContent());
        }

        // Tool call id tokens
        if (message.getToolCallId() != null) {
            tokens += countTokens(message.getToolCallId());
        }

        // Tool calls (assistant message with function calls)
        if (message.getToolCalls() != null) {
            for (ChatRequest.ToolCall tc : message.getToolCalls()) {
                tokens += estimateToolCall(tc);
            }
        }

        return tokens;
    }

    private int estimateToolCall(ChatRequest.ToolCall toolCall) {
        int tokens = 7; // tool_call structure overhead
        if (toolCall.getId() != null) {
            tokens += countTokens(toolCall.getId());
        }
        if (toolCall.getFunction() != null) {
            if (toolCall.getFunction().getName() != null) {
                tokens += countTokens(toolCall.getFunction().getName());
            }
            if (toolCall.getFunction().getArguments() != null) {
                tokens += countTokens(toolCall.getFunction().getArguments());
            }
        }
        return tokens;
    }

    /**
     * 估算 tool definitions 的 token 数
     */
    public int estimateToolDefinitions(List<ChatRequest.ToolDefinition> tools) {
        int total = 0;
        for (ChatRequest.ToolDefinition tool : tools) {
            total += estimateToolDefinition(tool);
        }
        // Format overhead for the tools array
        total += 12;
        return total;
    }

    private int estimateToolDefinition(ChatRequest.ToolDefinition tool) {
        int tokens = 7; // structure overhead
        if (tool.getType() != null) {
            tokens += countTokens(tool.getType());
        }
        if (tool.getFunction() != null) {
            ChatRequest.Function fn = tool.getFunction();
            if (fn.getName() != null) {
                tokens += countTokens(fn.getName());
            }
            if (fn.getDescription() != null) {
                tokens += countTokens(fn.getDescription());
            }
            if (fn.getParameters() != null) {
                try {
                    String json = objectMapper.writeValueAsString(fn.getParameters());
                    tokens += countTokens(json);
                } catch (Exception e) {
                    // fallback: rough estimate
                    tokens += fn.getParameters().size() * 20;
                }
            }
        }
        return tokens;
    }

    private static final int IMAGE_TOKEN_ESTIMATE = 1000;

    private int countImagePartTokens(Object content) {
        if (!(content instanceof List<?> list)) {
            return 0;
        }
        int images = 0;
        for (Object item : list) {
            if (item instanceof ChatRequest.ContentPart part) {
                if ("image_url".equals(part.getType())) {
                    images++;
                }
            } else if (item instanceof java.util.Map<?, ?> map) {
                if ("image_url".equals(map.get("type"))) {
                    images++;
                }
            }
        }
        return images * IMAGE_TOKEN_ESTIMATE;
    }

    /**
     * Extract text from content (String or List<ContentPart>)
     */
    public static String contentToString(Object content) {
        if (content == null) return "";
        if (content instanceof String s) return s;
        if (content instanceof List<?> list) {
            StringBuilder sb = new StringBuilder();
            for (Object item : list) {
                if (item instanceof ChatRequest.ContentPart part) {
                    if ("text".equals(part.getType()) && part.getText() != null) {
                        sb.append(part.getText());
                    }
                } else if (item instanceof java.util.Map<?, ?> map) {
                    Object type = map.get("type");
                    if ("text".equals(type)) {
                        Object text = map.get("text");
                        if (text != null) sb.append(text);
                    }
                }
            }
            return sb.toString();
        }
        return content.toString();
    }

    /**
     * UTF-8 字节粗算：{@code ceil(bytes / 4)}。
     */
    public int countTokens(String text) {
        if (text == null || text.isEmpty()) {
            return 0;
        }
        int bytes = text.getBytes(StandardCharsets.UTF_8).length;
        return (bytes + 3) / 4;
    }
}

File: backend-ts/src/harness/core/token-estimator.ts
import type { ChatMessage, ChatRequest, ContentPart, ToolCall, ToolDefinition } from '../llm/chat-request.js';

const MESSAGE_OVERHEAD_TOKENS = 4;
const IMAGE_TOKEN_ESTIMATE = 1000;

export class TokenEstimator {
  estimateRequestTokens(request: ChatRequest): number {
    let total = 0;
    if (request.messages) total += this.estimateMessages(request.messages);
    if (request.tools && request.tools.length > 0) total += this.estimateToolDefinitions(request.tools);
    return total;
  }

  estimateMessages(messages: ChatMessage[]): number {
    let total = 0;
    for (const msg of messages) total += this.estimateMessage(msg);
    return total;
  }

  estimateMessage(message: ChatMessage): number {
    let tokens = MESSAGE_OVERHEAD_TOKENS;
    if (message.role) tokens += this.countTokens(message.role);
    if (message.content != null) {
      tokens += this.countTokens(TokenEstimator.contentToString(message.content));
      tokens += this.countImagePartTokens(message.content);
    }
    if (message.toolCallId) tokens += this.countTokens(message.toolCallId);
    if (message.toolCalls) {
      for (const tc of message.toolCalls) tokens += this.estimateToolCall(tc);
    }
    return tokens;
  }

  private estimateToolCall(toolCall: ToolCall): number {
    let tokens = 7;
    if (toolCall.id) tokens += this.countTokens(toolCall.id);
    if (toolCall.function) {
      if (toolCall.function.name) tokens += this.countTokens(toolCall.function.name);
      if (toolCall.function.arguments) tokens += this.countTokens(toolCall.function.arguments);
    }
    return tokens;
  }

  estimateToolDefinitions(tools: ToolDefinition[]): number {
    let total = 0;
    for (const tool of tools) total += this.estimateToolDefinition(tool);
    total += 12;
    return total;
  }

  private estimateToolDefinition(tool: ToolDefinition): number {
    let tokens = 7;
    if (tool.type) tokens += this.countTokens(tool.type);
    if (tool.function) {
      const fn = tool.function;
      if (fn.name) tokens += this.countTokens(fn.name);
      if (fn.description) tokens += this.countTokens(fn.description);
      if (fn.parameters) {
        try {
          tokens += this.countTokens(JSON.stringify(fn.parameters));
        } catch {
          tokens += Object.keys(fn.parameters).length * 20;
        }
      }
    }
    return tokens;
  }

  private countImagePartTokens(content: unknown): number {
    if (!Array.isArray(content)) return 0;
    let images = 0;
    for (const item of content) {
      if (item && typeof item === 'object') {
        const type = (item as ContentPart).type ?? (item as Record<string, unknown>).type;
        if (type === 'image_url') images++;
      }
    }
    return images * IMAGE_TOKEN_ESTIMATE;
  }

  static contentToString(content: unknown): string {
    if (content == null) return '';
    if (typeof content === 'string') return content;
    if (Array.isArray(content)) {
      let sb = '';
      for (const item of content) {
        if (item && typeof item === 'object') {
          const o = item as Record<string, unknown>;
          if (o.type === 'text' && o.text != null) sb += String(o.text);
        }
      }
      return sb;
    }
    return String(content);
  }

  countTokens(text: string | null | undefined): number {
    if (text == null || text === '') return 0;
    const bytes = Buffer.byteLength(text, 'utf8');
    return Math.floor((bytes + 3) / 4);
  }
}
